Rich scala stubs

aljoscha opened the pull request:
Change Operators to accept stubs with open/close etc. You might call them rich stubs, but in the code it's just MapStub, FlatMapStub, FilterStub, ReduceStub, etc. They derive from MapStub etc. (the java one), so you can do things like getting the iteration context or what you might have...

This allows something like:

``` scala
.map( new MapStub[(String, Int), (String, Int)] {
  override def open(config: Configuration) = {
    println("Opening up this badboy.")
  }
  override def map(in: (String, Int)) = {
    println("IN RICH MAPPER: " + in)
    val (w, c) = in
    (w, c + 1)
  }
})
```

and likewise for flatMap and filter, for example:

``` scala
.map( new FlatMapStub[(String, Int), (String, Int)] {
  override def open(config: Configuration) = {
    println("Opening up this badboy.")
  }
  override def flatMap(in: (String, Int)) = {
    println("IN RICH MAPPER: " + in)
    in
  }
})
```

If people like this I will also make this available for all the other operators.

aljoscha committed:
Change map operator to allow a map stub with open/close etc

rmetzger commented on pact-scala/pact-scala-examples/src/main/scala/eu/stratosphere/scala/examples/grabbag/Grabbag.scala:
```diff
@@ -62,12 +70,22 @@ object Main1 extends Serializable {
+          println("Opening up this badboy.")
+        }
+        override def map(in: (String, Int)) = {
+          println("IN RICH MAPPER: " + in)
```
this looks like debugging code. Is it intended to have this in the master?

aljoscha replied:
Have you seen the rest of the file? :smile:

rmetzger replied:
Okay, I took a look :poop:
Why is it in the code then?

aljoscha replied:
Because someone merged it and no one complained so far... :smirk:

fhueske commented:
I think we need these rich interfaces. Otherwise, the Scala API is less expressive than the Java API which should not be the case.

StephanEwen commented on pact-scala/pact-scala-core/src/main/scala/eu/stratosphere/scala/stubs/MapStub.scala:
```diff
@@ -0,0 +1,71 @@
+import eu.stratosphere.pact.common.`type`.PactRecord
+
+abstract class MapStubBase[In: UDT, Out: UDT] extends JMapStub with Serializable with Function1[In, Out] {
+  val inputUDT: UDT[In] = implicitly[UDT[In]]
```
Does it make sense to make these private, to an implementing user does not mess with them?

StephanEwen commented on pact-scala/pact-scala-core/src/main/scala/eu/stratosphere/scala/stubs/MapStub.scala:
```diff
@@ -0,0 +1,71 @@
+  val outputUDT: UDT[Out] = implicitly[UDT[Out]]
+  lazy val udf: UDF1[In, Out] = new UDF1(inputUDT, outputUDT)
+
+  lazy val deserializer: UDTSerializer[In] = udf.getInputDeserializer
```
I guess you use a lazy val here to make sure it is not instantiated before being shipped and in its proper context.

We should replace this through some other mechanism at some point, because these vals are very frequently accessed. When mapping to generic records instead of pact record.

aljoscha replied:
First I did not use lazy vals and did the initialization in the open method. But it can happen that a user overrides open() and forgets to call super.open(). In that case the code will fail with null pointer exceptions. Even I forget this at first and was wondering whether I had made a mistake.

StephanEwen replied:
Yes, it is a valid workaround for the time. Is it possible to make them
private nonetheless?

On Wed, Dec 4, 2013 at 4:58 PM, Aljoscha Krettek
notifications@github.comwrote:

> In
> pact-scala/pact-scala-core/src/main/scala/eu/stratosphere/scala/stubs/MapStub.scala:
> 
> > @@ -0,0 +1,71 @@
> > +package eu.stratosphere.scala.stubs
> > +
> > +import eu.stratosphere.scala.analysis.{UDTSerializer, UDT, UDF1}
> > +import eu.stratosphere.pact.common.stubs.{Collector, MapStub => JMapStub}
> > +import eu.stratosphere.nephele.configuration.Configuration
> > +import eu.stratosphere.pact.common.`type`.PactRecord
> > +
> > +abstract class MapStubBase[In: UDT, Out: UDT] extends JMapStub with Serializable with Function1[In, Out] {
> > -  val inputUDT: UDT[In] = implicitly[UDT[In]]
> > -  val outputUDT: UDT[Out] = implicitly[UDT[Out]]
> > -  lazy val udf: UDF1[In, Out] = new UDF1(inputUDT, outputUDT)
> >   +
> > -  lazy val deserializer: UDTSerializer[In] = udf.getInputDeserializer
> 
> First I did not use lazy vals and did the initialization in the open
> method. But it can happen that a user overrides open() and forgets to call
> super.open(). In that case the code will fail with null pointer exceptions.
> Even I forget this at first and was wondering whether I had made a mistake.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/stratosphere/stratosphere/pull/319/files#r8099340
> .

aljoscha replied:
I will make them protected, because I use them for generated MapStubs.

aljoscha committed:
Change reduce operator to allow a reduce stub with open/close etc

aljoscha committed:
Change cross operator to allow a cross stub with open/close etc

aljoscha committed:
Make udf in scala MapStub and ReduceStub non-lazy

aljoscha committed:
Move DeserializingIterator to stubs package, clean-up imports

aljoscha committed:
Change join operator to allow a join stub with open/close etc

aljoscha committed:
Change cogroup operator to allow a cogroup stub with open/close etc

aljoscha committed:
Make all scala rich stubs have apply method instead of specific name

Before, MapStub required the user to specify a map() method, now they
specify an apply() method. This is more in line with how Function1[] and
so one works and also the same across all operators.

aljoscha commented:
This can now be merged, all operators have rich stubs.

StephanEwen commented:
Merged in c7febc1ff1c7d54313da29a07d2eaaf469aab252
